# API

* serializar  
        dict/list -> json

* deserializacion  
        json -> diccionario/lista

In [1]:
# diccionario

dic_ejemplo={'nombre': 'maria',
             'apellidos':{'apellido1':'perez', 'apellido2':'garcia'},
            'edad':45,
            'geoloc':(45.9, 18.9),
            'mascotas':[{'perro':'chucky'},{'gato':'doraemon'}]
            }

In [2]:
type(dic_ejemplo)

dict

In [3]:
dic_ejemplo

{'nombre': 'maria',
 'apellidos': {'apellido1': 'perez', 'apellido2': 'garcia'},
 'edad': 45,
 'geoloc': (45.9, 18.9),
 'mascotas': [{'perro': 'chucky'}, {'gato': 'doraemon'}]}

In [4]:
dic_ejemplo.keys()

dict_keys(['nombre', 'apellidos', 'edad', 'geoloc', 'mascotas'])

In [5]:
dic_ejemplo.values()

dict_values(['maria', {'apellido1': 'perez', 'apellido2': 'garcia'}, 45, (45.9, 18.9), [{'perro': 'chucky'}, {'gato': 'doraemon'}]])

In [6]:
lista_ejemplo=['string', 15,8,9, [4,5,6], None]

In [7]:
type(lista_ejemplo)

list

### SERIALIZACION A JSON  
##### javascript object anotation

In [8]:
import json

In [9]:
json_dict=json.dumps(dic_ejemplo)

In [10]:
type(json_dict)

str

In [11]:
json_dict

'{"nombre": "maria", "apellidos": {"apellido1": "perez", "apellido2": "garcia"}, "edad": 45, "geoloc": [45.9, 18.9], "mascotas": [{"perro": "chucky"}, {"gato": "doraemon"}]}'

In [12]:
json_list=json.dumps(lista_ejemplo)

In [13]:
type(json_list)

#es una cadena

str

In [14]:
json_list[4:25]
#devuelve de la posoicion 5 a la 24 de la cadena

'ring", 15, 8, 9, [4, '

### DESERIALIZACION DE JSON

In [15]:
json_to_dict=json.loads(json_dict)

In [16]:
type(json_to_dict)

#vuelve a diccionario

dict

In [17]:
json_to_dict

{'nombre': 'maria',
 'apellidos': {'apellido1': 'perez', 'apellido2': 'garcia'},
 'edad': 45,
 'geoloc': [45.9, 18.9],
 'mascotas': [{'perro': 'chucky'}, {'gato': 'doraemon'}]}

In [18]:
json_to_list=json.loads(json_list)

In [19]:
type(json_to_list)

list

### CONEXION CON API/DB/Servidores

In [20]:
# libreria con metodos GET y POST
import requests
import urllib
import pandas as pd
import numpy as np


In [39]:
# estructura solicitud request

api_url='http://api.postcodes.io/postcodes/'
zipcode='NE30 1DP'

# response recibe la información solcitada con GET (requests.get())
r=requests.get(api_url+zipcode)

In [40]:
if r.status_code in range(200,300):
    print("request OK")
    

request OK


In [41]:
# Cabcera del GET
r.headers

{'Date': 'Sat, 10 Jun 2023 20:24:30 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'x-gnu': 'Michael J Blanchard', 'access-control-allow-origin': '*', 'etag': 'W/"3b6-BrmR8ce2+HDRdMnRSvrjHypVDJE"', 'CF-Cache-Status': 'HIT', 'Age': '8976', 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=ltK1%2BWWdBMHDG1z6li%2Fofn4KRpLsyqzJpszKwAw13yahzyHxchTSb%2Ff5qYgHqBcb7BNosi627bxbapsqYRU1ecMt5dOAW0C%2B1itoWBYD2Fc3WrAqy8%2BfFqVoIbBYNuzp5Uo%3D"}],"group":"cf-nel","max_age":604800}', 'NEL': '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}', 'Vary': 'Accept-Encoding', 'Server': 'cloudflare', 'CF-RAY': '7d5456972e836689-MAD', 'Content-Encoding': 'gzip'}

In [42]:
# Información de la solicitud
r.text

'{"status":200,"result":{"postcode":"NE30 1DP","quality":1,"eastings":435958,"northings":568671,"country":"England","nhs_ha":"North East","longitude":-1.439269,"latitude":55.011303,"european_electoral_region":"North East","primary_care_trust":"North Tyneside","region":"North East","lsoa":"North Tyneside 016C","msoa":"North Tyneside 016","incode":"1DP","outcode":"NE30","parliamentary_constituency":"Tynemouth","admin_district":"North Tyneside","parish":"North Tyneside, unparished area","admin_county":null,"date_of_introduction":"201303","admin_ward":"Tynemouth","ced":null,"ccg":"NHS North East and North Cumbria","nuts":"Tyneside","pfa":"Northumbria","codes":{"admin_district":"E08000022","admin_county":"E99999999","admin_ward":"E05001130","parish":"E43000176","parliamentary_constituency":"E14001006","ccg":"E38000127","ccg_id":"99C","ced":"E99999999","nuts":"TLC22","lsoa":"E01008561","msoa":"E02001753","lau2":"E08000022","pfa":"E23000007"}}}'

* estructura de la información  
    > status y result

'{"status":200,"result":{"postcode":"NE30 1DP","quality":1,"eastings":435958,"northings":568671,"country":"England","nhs_ha":"North East","longitude":-1.439269,"latitude":55.011303,"european_electoral_region":"North East","primary_care_trust":"North Tyneside","region":"North East","lsoa":"North Tyneside 016C","msoa":"North Tyneside 016","incode":"1DP","outcode":"NE30","parliamentary_constituency":"Tynemouth","admin_district":"North Tyneside","parish":"North Tyneside, unparished area","admin_county":null,"date_of_introduction":"201303","admin_ward":"Tynemouth","ced":null,"ccg":"NHS North East and North Cumbria","nuts":"Tyneside","pfa":"Northumbria","codes":{"admin_district":"E08000022","admin_county":"E99999999","admin_ward":"E05001130","parish":"E43000176","parliamentary_constituency":"E14001006","ccg":"E38000127","ccg_id":"99C","ced":"E99999999","nuts":"TLC22","lsoa":"E01008561","msoa":"E02001753","lau2":"E08000022","pfa":"E23000007"}}}'

In [43]:
r.headers["Date"]

'Sat, 10 Jun 2023 20:24:30 GMT'

In [44]:
r.url

'http://api.postcodes.io/postcodes/NE30%201DP'

In [45]:
r.content
#tipo byte (b' al principio)

b'{"status":200,"result":{"postcode":"NE30 1DP","quality":1,"eastings":435958,"northings":568671,"country":"England","nhs_ha":"North East","longitude":-1.439269,"latitude":55.011303,"european_electoral_region":"North East","primary_care_trust":"North Tyneside","region":"North East","lsoa":"North Tyneside 016C","msoa":"North Tyneside 016","incode":"1DP","outcode":"NE30","parliamentary_constituency":"Tynemouth","admin_district":"North Tyneside","parish":"North Tyneside, unparished area","admin_county":null,"date_of_introduction":"201303","admin_ward":"Tynemouth","ced":null,"ccg":"NHS North East and North Cumbria","nuts":"Tyneside","pfa":"Northumbria","codes":{"admin_district":"E08000022","admin_county":"E99999999","admin_ward":"E05001130","parish":"E43000176","parliamentary_constituency":"E14001006","ccg":"E38000127","ccg_id":"99C","ced":"E99999999","nuts":"TLC22","lsoa":"E01008561","msoa":"E02001753","lau2":"E08000022","pfa":"E23000007"}}}'

In [46]:
r.text
#tipo string. Hay que serializarlo

'{"status":200,"result":{"postcode":"NE30 1DP","quality":1,"eastings":435958,"northings":568671,"country":"England","nhs_ha":"North East","longitude":-1.439269,"latitude":55.011303,"european_electoral_region":"North East","primary_care_trust":"North Tyneside","region":"North East","lsoa":"North Tyneside 016C","msoa":"North Tyneside 016","incode":"1DP","outcode":"NE30","parliamentary_constituency":"Tynemouth","admin_district":"North Tyneside","parish":"North Tyneside, unparished area","admin_county":null,"date_of_introduction":"201303","admin_ward":"Tynemouth","ced":null,"ccg":"NHS North East and North Cumbria","nuts":"Tyneside","pfa":"Northumbria","codes":{"admin_district":"E08000022","admin_county":"E99999999","admin_ward":"E05001130","parish":"E43000176","parliamentary_constituency":"E14001006","ccg":"E38000127","ccg_id":"99C","ced":"E99999999","nuts":"TLC22","lsoa":"E01008561","msoa":"E02001753","lau2":"E08000022","pfa":"E23000007"}}}'

In [57]:
contenido=json.loads(r.text)
contenido

{'status': 200,
 'result': {'postcode': 'NE30 1DP',
  'quality': 1,
  'eastings': 435958,
  'northings': 568671,
  'country': 'England',
  'nhs_ha': 'North East',
  'longitude': -1.439269,
  'latitude': 55.011303,
  'european_electoral_region': 'North East',
  'primary_care_trust': 'North Tyneside',
  'region': 'North East',
  'lsoa': 'North Tyneside 016C',
  'msoa': 'North Tyneside 016',
  'incode': '1DP',
  'outcode': 'NE30',
  'parliamentary_constituency': 'Tynemouth',
  'admin_district': 'North Tyneside',
  'parish': 'North Tyneside, unparished area',
  'admin_county': None,
  'date_of_introduction': '201303',
  'admin_ward': 'Tynemouth',
  'ced': None,
  'ccg': 'NHS North East and North Cumbria',
  'nuts': 'Tyneside',
  'pfa': 'Northumbria',
  'codes': {'admin_district': 'E08000022',
   'admin_county': 'E99999999',
   'admin_ward': 'E05001130',
   'parish': 'E43000176',
   'parliamentary_constituency': 'E14001006',
   'ccg': 'E38000127',
   'ccg_id': '99C',
   'ced': 'E99999999',


In [48]:
contenido.keys()

dict_keys(['status', 'result'])

In [49]:
contenido['status']

200

In [50]:
contenido['result']

{'postcode': 'NE30 1DP',
 'quality': 1,
 'eastings': 435958,
 'northings': 568671,
 'country': 'England',
 'nhs_ha': 'North East',
 'longitude': -1.439269,
 'latitude': 55.011303,
 'european_electoral_region': 'North East',
 'primary_care_trust': 'North Tyneside',
 'region': 'North East',
 'lsoa': 'North Tyneside 016C',
 'msoa': 'North Tyneside 016',
 'incode': '1DP',
 'outcode': 'NE30',
 'parliamentary_constituency': 'Tynemouth',
 'admin_district': 'North Tyneside',
 'parish': 'North Tyneside, unparished area',
 'admin_county': None,
 'date_of_introduction': '201303',
 'admin_ward': 'Tynemouth',
 'ced': None,
 'ccg': 'NHS North East and North Cumbria',
 'nuts': 'Tyneside',
 'pfa': 'Northumbria',
 'codes': {'admin_district': 'E08000022',
  'admin_county': 'E99999999',
  'admin_ward': 'E05001130',
  'parish': 'E43000176',
  'parliamentary_constituency': 'E14001006',
  'ccg': 'E38000127',
  'ccg_id': '99C',
  'ced': 'E99999999',
  'nuts': 'TLC22',
  'lsoa': 'E01008561',
  'msoa': 'E02001

In [51]:
contenido['result'].keys()

dict_keys(['postcode', 'quality', 'eastings', 'northings', 'country', 'nhs_ha', 'longitude', 'latitude', 'european_electoral_region', 'primary_care_trust', 'region', 'lsoa', 'msoa', 'incode', 'outcode', 'parliamentary_constituency', 'admin_district', 'parish', 'admin_county', 'date_of_introduction', 'admin_ward', 'ced', 'ccg', 'nuts', 'pfa', 'codes'])

In [52]:
pd.DataFrame(contenido['result']).head(2)

postcode  quality  eastings  northings  country      nhs_ha   
admin_county    NE30 1DP        1    435958     568671  England  North East  \
admin_district  NE30 1DP        1    435958     568671  England  North East   

                longitude   latitude european_electoral_region   
admin_county    -1.439269  55.011303                North East  \
admin_district  -1.439269  55.011303                North East   

               primary_care_trust  ...  admin_district   
admin_county       North Tyneside  ...  North Tyneside  \
admin_district     North Tyneside  ...  North Tyneside   

                                         parish admin_county   
admin_county    North Tyneside, unparished area         None  \
admin_district  North Tyneside, unparished area         None   

               date_of_introduction admin_ward   ced   
admin_county                 201303  Tynemouth  None  \
admin_district               201303  Tynemouth  None   

                                             ccg      nuts          pfa   
admin_county    NHS North East and North Cumbria  Tyneside  Northumbria  \
admin_district  NHS North East and North Cumbria  Tyneside  Northumbria   

                    codes  
admin_county    E99999999  
admin_district  E08000022  

[2 rows x 26 columns]

In [53]:
contenido['result']['region']

'North East'

In [54]:
# sacar del diccionario solo los resultados que queramos y almacenarlos en un diccionario

new_dict={
    'zipcode':contenido['result']['postcode'],
    'region':contenido['result']['region']
}

In [55]:
new_dict

{'zipcode': 'NE30 1DP', 'region': 'North East'}

In [56]:
pd.DataFrame(new_dict, index=[0])

zipcode      region
0  NE30 1DP  North East